timeline/src/Data/Timeline/Internal.hs

```diff
@@ -0,0 +1,299 @@
+    value :: a
+  }
+  deriving stock (Eq, Functor)
+
+type Getter s a = forall f. (Contravariant f) => (a -> f a) -> s -> f s
+
+recordEffectiveFrom :: Getter (Record a) UTCTime
+recordEffectiveFrom f r = contramap getEffectiveFrom (f (getEffectiveFrom r))
+  where
+    getEffectiveFrom :: Record a -> UTCTime
+    getEffectiveFrom Record {effectiveFrom} = effectiveFrom
+
+recordEffectiveTo :: Getter (Record a) (Maybe UTCTime)
+recordEffectiveTo f r = contramap getEffectiveTo (f (getEffectiveTo r))
+  where
+    getEffectiveTo :: Record a -> Maybe UTCTime
+    getEffectiveTo Record {effectiveTo} = effectiveTo
+
+recordValue :: Getter (Record a) a
+recordValue f r = contramap getValue (f (getValue r))
+  where
+    getValue :: Record a -> a
+    getValue Record {value} = value
+
+-- | 'makeRecord' returns 'Nothing' if @effectiveTo@ is not greater than @effectiveFrom@
+makeRecord ::
+  -- | effective from
+  UTCTime ->
+  -- | optional effective to
+  Maybe UTCTime ->
+  -- | value
+  a ->
+  Maybe (Record a)
+-- can't write a makeRecordTH because UTCTime has no Lift instance
+makeRecord effectiveFrom effectiveTo value =
+  if maybe False (effectiveFrom >=) effectiveTo
+    then Nothing
+    else Just Record {..}
+
+instance (TH.Lift a) => TH.Lift (Record a) where
+  liftTyped Record {..} =
+    [||
+    Record
+      (unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime effectiveFrom))
+      (fmap unLiftUTCTime $$(TH.liftTyped $ LiftUTCTime <$> effectiveTo))
+      $$(TH.liftTyped value)
+    ||]
+
+newtype LiftUTCTime = LiftUTCTime UTCTime
+  deriving stock (Generic)
+
+unLiftUTCTime :: LiftUTCTime -> UTCTime
+unLiftUTCTime (LiftUTCTime t) = t
+
+instance TH.Lift LiftUTCTime where
+  liftTyped (LiftUTCTime t) = [||LiftUTCTime (read $$(TH.liftTyped (show t)))||]
+
+instance (Show a) => Show (Record a) where
+  show Record {..} = show effectiveFrom <> " ~ " <> show effectiveTo <> ": " <> show value
+
+newtype Overlaps a = Overlaps {groups :: NonEmpty (OverlapGroup a)}
+  deriving newtype (Semigroup)
+  deriving stock (Eq, Generic)
+
+instance (Show a) => Show (Overlaps a) where
+  show Overlaps {groups} =
+    "Here are " <> show (length groups) <> " group(s) of overlapping records\n"
+      ++ sep
+      ++ intercalate sep (show <$> NonEmpty.toList groups)
+      ++ sep
+    where
+      sep = "--------------------\n"
+
+data OverlapGroup a = OverlapGroup (Record a) (Record a) [Record a]
+  deriving stock (Eq, Generic)
+
+instance (Show a) => Show (OverlapGroup a) where
+  show = unlines . fmap show . unpackOverlapGroup
+
+unpackOverlapGroup :: OverlapGroup a -> [Record a]
+unpackOverlapGroup (OverlapGroup r1 r2 records) = r1 : r2 : records
+
+-- Build a 'Timeline' from a list of 'Record's.
+--
+-- For any time, there could be zero, one, or more values, according to the input. No other condition
+-- is possible. We have taken account the "zero" case by wrapping the result in 'Maybe', so the only
+-- possible error is 'Overlaps'.
```

Comment:
Mention that we have a `traversable` instance, so you can go `Timeline (Maybe a)` -> `Maybe (Timeline a)`?